### DISCUSSION TITLE
(Bug?) Getting an atom value after awaiting async function lead to unintended behavior in derived atom

### DISCUSSION THREAD

**Discussion by devnomic** [pmndrs/jotai#1905] [General]
Getting an atom after awaiting async function lead to unintended behavior when you using the derived atom.
Sometime it works sometime it's not. 
I think its a bug, or is it working as intended? @dai-shi 

````
const counterAtom = atom(0);
const valueAtom = atom(async (get) => {
  try {
    // Do some fetching from server or getting value from another async atom
    await new Promise((resolve) => setTimeout(resolve, 1000)); // 1 sec

    // Get the counter after promise
    // will result in unintended behavior (bugs) in deriving atom

    const counter = get(counterAtom);
    return 10 + counter;
  } catch {
    return 0;
  }
});

// Using this derived atom will cause unintended behavior.
const valueTextAtom = atom(async (get) => {
  const value = await get(valueAtom);
  return value.toFixed(2);
});
````

Component
````
function CounterButton() {
  const [counter, setCounter] = useAtom(counterAtom);
  return (
    <button onClick={() => setCounter((prev) => prev + 1)}>
      Counter: {counter}
    </button>
  );
}

function ValueText() {  
  // useAtomValue(valueTextAtom) will result in unintended behavior
  // useAtomValue(valueAtom) everything works as intended.
  const valueText = useAtomValue(valueTextAtom);
  return <div>{valueText}</div>;
}

export default function App() {
  return (
    <div className="App">
      <h1>Hello CodeSandbox</h1>
      <h2>Start editing to see some magic happen!</h2>
      <Suspense fallback={<div>Please wait</div>}>
        <ValueText />
      </Suspense>
      <CounterButton />
    </div>
  );
}
````

**Repro:**
https://codesandbox.io/s/infallible-tamas-tww3en?file=/src/App.tsx

**Visual screen recording:**
![Screen Recording 2023-04-30 at 22 26 13](https://user-images.githubusercontent.com/101681592/235361692-6a3d3ce0-e5f9-44de-8c38-6cdc76df4d03.gif)

**Comment by dai-shi** [comment]
Thanks for reporting. It seems like a bug.
Hoped the current #1851 fixes it, but it doesn't.
v1 works fine, so it's a regression in v2.
I wonder if we can make a store only test in `./tests/vanilla/store.test.tsx`.

**Reply by dai-shi** [reply]
Tried it in https://github.com/pmndrs/jotai/pull/1851/commits/11b515e761be81c86e234bbe890d234012c119ac, but it's passing. Can anyone double check it?

**Reply by arjunvegda** [reply]
~Could it be a stale atom subscription? 🤔 I can try pushing a failing test.~ Actually, I'm not sure.

Subbing to the store causes it to fail though.
```diff
it('should update async atom with deps after await (#1905)', async () => {
  const countAtom = atom(0)
  const resolve: (() => void)[] = []
  const delayedAtom = atom(async (get) => {
    await new Promise<void>((r) => resolve.push(r))
    const count = get(countAtom)
    return count
  })
  const derivedAtom = atom(async (get) => {
    const count = await get(delayedAtom)
    return count
  })

  const store = createStore()
+  const l = jest.fn()
+  const unsub = store.sub(derivedAtom, l)
  const promise0 = store.get(derivedAtom)
  resolve.splice(0).forEach((fn) => fn())
  expect(await promise0).toBe(0)

  store.set(countAtom, 1)
  const promise1 = store.get(derivedAtom)
  resolve.splice(0).forEach((fn) => fn())
  expect(await promise1).toBe(1)
  store.set(countAtom, 2)
  const promise2 = store.get(derivedAtom)
  resolve.splice(0).forEach((fn) => fn())
  expect(await promise2).toBe(2)
  store.set(countAtom, 3)
  const promise3 = store.get(derivedAtom)
  resolve.splice(0).forEach((fn) => fn())
  expect(await promise3).toBe(3)
+  expect(l).toHaveBeenCalledTimes(3)
+  unsub()
})
```

**Reply by devnomic** [reply]
i tried in codesandbox build, not working
https://codesandbox.io/s/react-typescript-forked-0wl31i?file=/src/App.tsx

**Reply by dai-shi** [reply]
> `const unsub = store.sub(derivedAtom, l)`

Oh, nice catch. It should be it.

**Reply by dai-shi** [reply]
https://github.com/pmndrs/jotai/pull/1851/commits/a7aa7535b151db1741cf3089bc99c807a418f2c3 Now it fails.

**Reply by dai-shi** [reply]
@devnomic 
#1851 fixes it: https://codesandbox.io/s/naughty-jones-9zi2u7

**Reply by devnomic** [reply]
It seems working very well! hopefully this get merged soon as this quite important fix imo.

**Reply by dai-shi** [reply]
yeah.

**Comment by himself65** [comment]
I found a hotfix for the current version, add a `sub` on that atom

```ts
rootStore.sub(workspaceIdsAtom, () => {})
```

**Reply by himself65** [reply]
https://github.com/Himself65/affine-demo/commit/df1bfcfb48e9226b30eaaed48595cd968e1c7b1e#diff-2717c7fb31c2070dd96f07394997809bc9be7f3fff03ab638db696cc480f39b6R21-R25
